Modified templates and customtags for study and animal tables

fkang-pu opened the pull request:
<!-- markdownlint-disable-next-line first-line-heading -->
## Summary Change Description

- Modified code for customtag filter "obj_hyperlink", allowing comma separated HTML string to display content with or without line breaks.

- Added "total infusate" to the DataFrame for study list.

- Made changes to study and animal tables for easier read, while keeping search/filter/download features.

- Study list page
    - hide "infusate" column as default, but can still toggle to show the column.
    - added "total infusate" column,  including a feature to show/hide infusate details by clicking the arrow icon.  
    - changed column width for study description.  
    
## Affected Issue Numbers
- Resolved #612
- Partially addresses #613
- Partially addresses #614
- Partially addresses #615
- Partially addresses #616
- Notes:
   Several Issues above are marked as partially addressed, since no action was made to protocol and peakgroupset tables yet.

## Code Review Notes
- Bootstrap-table plugin has attributes to control column width, but the settings don't work for responsive table with a large number of columns. Adding column width control improved table layout for study list table, but it didn't show effect for animal list table (19 columns). Need to consider other options for further improvements.

- I added change logs to version 2.0.2. My assumption is we wanted to push out the template changes before the Lab meeting next week. Got one linter error for changelogs.md. Any suggestion on how to handle the headings with new release?
   error: Multiple headings with the same content [Context: "### Changed"]

## Additional Code Review Notes (added on Jan. 31, 2023)
- Added a new css file: `bootstrap_table_cus1.css`
- Added a new JavaScript file: `setTableHeight.js`
- fixed `empty space` and `two scrollbars` issues and made improvements on `column width` for the following pages:
   - study list
   - study detail
   - study summary
   - animal list
   - animal detail
   -  sample list
   - compound detail
 - Added changes  to  `Unreleased` section of changelog.

## Checklist

- [x] Changes have been added to the *Unreleased* section in the [`changelog.md`](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/changelog.md).
- [ ] All issue requirements satisfied (or no linked issues)
- [ ] [Linting passes](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#linting).
- [ ] [Migrations created & committed *(or no model changes)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#migration-process)
- Tests
  - [ ] [Tests implemented *(or no code changes)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#test-implementation)
  - [ ] [All tests pass](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#quality-control)
  - [ ] All example load tests pass (remotely) *(or their failures predated and are unrelated to this branch)*

lparsons commented:
@fkang-pu To address the headings issue, I believe you can add to the `./.markdown-lint.yml` file. See [`md024` docs](https://github.com/DavidAnson/markdownlint/blob/main/doc/md024.md).

```yaml
MD024:
  siblings_only: true
```

hepcat72 commented on DataRepo/templates/DataRepo/includes/animal_list_stats_table.html:
```diff
@@ -23,12 +23,12 @@
             <tr>
                 <th data-filter-control="input" data-sortable="true" data-sorter="htmlSorter" data-field="Animal">Animal</th>
-                <th data-filter-control="input" data-sortable="true" data-sorter="htmlSorter" data-field="Studies">Studies</th>
+                <th data-filter-control="input" data-sortable="true" data-sorter="htmlSorter" data-width="7" data-width-unit="%" data-field="Studies">Studies</th>
```
Q. Is this width targeted toward getting the whole header in one line?  What happens in the study name is longer than the header?  Do the cells wrap or expand their width?

fkang-pu replied:
The width control is to make sure the "studies" column not too wide with long name(s) for studies. The content wraps in my dev. instance.

hepcat72 replied:
I guess what I meant was, was the specific width of 7% selected in order to not wrap the header?  Personally speaking, if it's possible, I would set a min/max and allow the header to wrap un less content pushes it out... but that's my aesthetic opinion, and I'm not known for my aesthetic choices...

fkang-pu replied:
There are 19 columns in total for animal list table, 4 are hidden as default. The average width for 15 columns would be 6.6%. That's why I set it to 7%. Honestly, I don't know if the width control works for a responsive table with current version of Bootstrap-table ([ref](https://bootstrap-table.com/docs/api/column-options/?#width)., [example](https://github.com/wenzhixin/bootstrap-table/issues/6526)). I can only tell the table display are better than current version in production, with the changes I made in this PR (obj_hyperlink tag and table attributes). I haven't found a way to set min/max. width for a responsive table with Bootstrap-table.

fkang-pu replied:
I am resolving this old conversation about `animal_list_stats` template, since I made improvements on handling column header and width.

hepcat72 commented on DataRepo/templates/DataRepo/includes/animal_list_stats_table.html:
```diff
@@ -63,7 +63,7 @@
                 </td>
                 <td>{{ i.compound_id_name_list|obj_hyperlink:"compound" }}</td>
-                <td>{{ i.tracer_id_name_list|obj_hyperlink:"tracer" }}</td>
+                <td>{{ i.tracer_id_name_list|obj_hyperlink:"tracer, newline" }}</td>
```
nb. If you need multiple arguments and the returned string is just going in the html output, you should register the `obj_hyperlink` tag as a `simple_tag`, not a `filter`.  Then you can pass multiple arguments without dealing with parsing multiple arguments out of a string.

fkang-pu replied:
I made changes to convert `obj_hyperlink ` to  a `simple_tag`. The code is simpler, and the lines to call this tag in templates are easier to follow. Thanks for the suggestion.

hepcat72 commented on DataRepo/templatetags/customtags.py:
```diff
@@ -146,18 +151,28 @@ def obj_hyperlink(id_name_list, obj):
-        '<div class="newlines">'
-        + format_html_join(
-            ",\n",
+    if len(args.split(",")) == 2 and args.split(",")[1].strip() == "newline":
```
nb. My previous comment about changing to `simple_tag` will make this moot, but if you don't do that, I would suggest processing all the args at the top of the method like:

```
def obj_hyperlink(id_name_list, str):
    args = str.split(",")
    obj = args[0]
    use_nls = (len(args) == 2 and args[1] == "newline")
```

hepcat72 reviewed (approved):
I'm approving, but I would prefer to see the suggested changes made.

In the future, it would help in terms of scope during review to create the issues first and link them in the PR description.

WRT scope, one issue that remains is that the table height in at least 1 of the edited files still exceeds my window size despite there being only a few rows in my sandbox database.

fkang-pu commented:
> WRT scope, one issue that remains is that the table height in at least 1 of the edited files still exceeds my window size despite there being only a few rows in my sandbox database.

@hepcat72 Could you let me know which table you had window size issue? I tested the display of each table with my 13'' laptop, but might have missed the issue.

hepcat72 commented:
> > WRT scope, one issue that remains is that the table height in at least 1 of the edited files still exceeds my window size despite there being only a few rows in my sandbox database.
> 
> @hepcat72 Could you let me know which table you had window size issue? I tested the display of each table with my 13'' laptop, but might have missed the issue.

I know it was the animal_list_stats_table.  I assume that was from clicking animals on the home page.  I don't have your branch loaded ATM, so I can't easily check.

fkang-pu commented:
> I know it was the animal_list_stats_table. I assume that was from clicking animals on the home page. I don't have your branch loaded ATM, so I can't easily check.

I checked and din't see the issue on my end.

hepcat72 commented:
It's opportune now. I'll switch branches and send a screenshot momentarily.

hepcat72 commented:
<img width="1696" alt="Animals - Screen Shot 2023-01-25 at 5 20 44 PM" src="https://user-images.githubusercontent.com/2300532/214706036-d4d24b27-a853-4b71-b833-d9bd5324301e.png">
<img width="1696" alt="Animals - Screen Shot 2023-01-25 at 5 20 57 PM" src="https://user-images.githubusercontent.com/2300532/214706099-5919023a-15ac-4964-af42-1e0b329365da.png">

lparsons commented:
> <img alt="Animals - Screen Shot 2023-01-25 at 5 20 44 PM" width="100" src="https://user-images.githubusercontent.com/2300532/214706036-d4d24b27-a853-4b71-b833-d9bd5324301e.png">
> 
> <img alt="Animals - Screen Shot 2023-01-25 at 5 20 57 PM" width="100" src="https://user-images.githubusercontent.com/2300532/214706099-5919023a-15ac-4964-af42-1e0b329365da.png">

The incorrect size of the table is my biggest issue with these pages atm, though I don't know that it is a new issue with this PR, I believe it's existed for some time.

fkang-pu commented:
> The incorrect size of the table is my biggest issue with these pages atm, though I don't know that it is a new issue with this PR, I believe it's existed for some time.

I pushed a commit to fix it. The animal list table uses the same template as that in compound detail. The issue is to display a table with total rows <10. I missed this table since I have more data in my local database.

hepcat72 commented:
> > The incorrect size of the table is my biggest issue with these pages atm, though I don't know that it is a new issue with this PR, I believe it's existed for some time.
> 
> I pushed a commit to fix it. The animal list table uses the same template as that in compound detail. The issue is to display a table with total rows <10. I missed this table since I have more data in my local database.

I'm curious.  If I had 11 rows in the table, wouldn't the "Showing..." thing still be off the bottom of the screen?  And if I had 200 rows, would the table height increase or stay a constant, but scrollable size?  If it's the latter, I would think that the position of the bottom edge would be better made to be relative to the bottom of the window...

fkang-pu commented:
> > > The incorrect size of the table is my biggest issue with these pages atm, though I don't know that it is a new issue with this PR, I believe it's existed for some time.
> > 
> > 
> > I pushed a commit to fix it. The animal list table uses the same template as that in compound detail. The issue is to display a table with total rows <10. I missed this table since I have more data in my local database.
> 
> I'm curious. If I had 11 rows in the table, wouldn't the "Showing..." thing still be off the bottom of the screen? And if I had 200 rows, would the table height increase or stay a constant, but scrollable size? If it's the latter, I would think that the position of the bottom edge would be better made to be relative to the bottom of the window...

There was an error in previous commit. I fixed it.  The aim is to have reasonable table height while keeping pagination and vertical scrolling features.  The table height control was added to several pages (animal list/detail, sample list), using JavaScript code implemented for compound detail page previously.  I found it's challenging to deal with a responsive table, since the column/row width and height vary greatly. I am open to suggestions.

lparsons reviewed (approved):
This looks like an improvement to me. I do think the CHANGELOG should reflect that these are unreleased until we decide to tag and release a new version.

lparsons reviewed (approved):
This looks great to me, the table heights are much better. I did notice that we might want to apply the same table height logic to:
* list of tissues
* list of peak group sets
* list of compounds
* list of infusates

Perhaps using some common template for these "list" pages?

fkang-pu commented:
> This looks great to me, the table heights are much better. I did notice that we might want to apply the same table height logic to:
> 
> * list of tissues
> * list of peak group sets
> * list of compounds
> * list of infusates
> 
> Perhaps using some common template for these "list" pages?

Plan to make improvements to the pages you listed above in a new branch/PR later this week.

hepcat72 commented:
> > I know it was the animal_list_stats_table. I assume that was from clicking animals on the home page. I don't have your branch loaded ATM, so I can't easily check.
> 
> I checked and din't see the issue on my end.

Regardless of the amount of data on each page, the controls under **almost every** table are off the bottom of the window for me.  I pulled your latest changes.  This is what I see currently:

<img width="1047" alt="Infusates - Screen Shot 2023-02-13 at 10 46 37 AM" src="https://user-images.githubusercontent.com/2300532/218505584-c5663f7d-fb03-4541-a3c3-677abd78d27d.png">
<img width="1047" alt="Infusates - Screen Shot 2023-02-13 at 10 46 24 AM" src="https://user-images.githubusercontent.com/2300532/218505601-75837125-e0b4-4d37-b566-97d8a42a29e1.png">
<img width="1047" alt="PeakGroup Sets - Screen Shot 2023-02-13 at 10 45 53 AM" src="https://user-images.githubusercontent.com/2300532/218505620-2039e457-3ca2-4fd5-b7c3-2982dce61704.png">
<img width="1047" alt="Tissues - Screen Shot 2023-02-13 at 10 44 54 AM png - Screen Shot 2023-02-13 at 10 45 08 AM" src="https://user-images.githubusercontent.com/2300532/218505652-65362ba8-50ea-42da-af7b-a0183ebcc79a.png">
<img width="1047" alt="studies - Screen Shot 2023-02-13 at 10 44 54 AM" src="https://user-images.githubusercontent.com/2300532/218505682-33750c83-2be2-4018-b23c-f55bfe4aca6b.png">

hepcat72 commented:
I only see the `Showing x of x rows` message on 1 page and the scrollbar blocks the 1 row shown.  None of any of the others have the pagination controls visible in the current window size.  The table height should be based on the size of the window (unless the content doesn't fill it up).

fkang-pu commented:
> I only see the `Showing x of x rows` message on 1 page and the scrollbar blocks the 1 row shown. None of any of the others have the pagination controls visible in the current window size. The table height should be based on the size of the window (unless the content doesn't fill it up).

@hepcat72 Thanks for your feedback. I'll spend a bit time to have a better understanding of Bootstrap-table plugin and address the issues in [PR 626](https://github.com/Princeton-LSI-ResearchComputing/tracebase/pull/626).  At least, the table height need to set properly when there are a few rows  or when the page size changes. I converted PR626 to a draft this morning.

hepcat72 commented:
> > I only see the `Showing x of x rows` message on 1 page and the scrollbar blocks the 1 row shown. None of any of the others have the pagination controls visible in the current window size. The table height should be based on the size of the window (unless the content doesn't fill it up).
> 
> @hepcat72 Thanks for your feedback. I'll spend a bit time to have a better understanding of Bootstrap-table plugin and address the issues in [PR 626](https://github.com/Princeton-LSI-ResearchComputing/tracebase/pull/626). At least, the table height need to set properly when there are a few rows or when the page size changes. I converted PR626 to a draft this morning.

Some things to consider - if you are setting the table height for a table with only a few rows:

- Account for space for a horizontal scrollbar (test by dragging your window to be narrow)
- Use a different (not the fancy static) header for those tables with only a few rows and perhaps in those instances, don't set a table height so that the table size is automatic

my 2¢

fkang-pu commented:
> > > I only see the `Showing x of x rows` message on 1 page and the scrollbar blocks the 1 row shown. None of any of the others have the pagination controls visible in the current window size. The table height should be based on the size of the window (unless the content doesn't fill it up).
> > 
> > 
> > @hepcat72 Thanks for your feedback. I'll spend a bit time to have a better understanding of Bootstrap-table plugin and address the issues in [PR 626](https://github.com/Princeton-LSI-ResearchComputing/tracebase/pull/626). At least, the table height need to set properly when there are a few rows or when the page size changes. I converted PR626 to a draft this morning.
> 
> Some things to consider - if you are setting the table height for a table with only a few rows:
> 
> * Account for space for a horizontal scrollbar (test by dragging your window to be narrow)
> * Use a different (not the fancy static) header for those tables with only a few rows and perhaps in those instances, don't set a table height so that the table size is automatic
> 
> my 2¢

That's was my solution before PR602 (not set table height for row <5). But it's more complex as row height varies a lot for summary tables (e.g. fixed table height is good for study list table with a few rows).  My current solution has some issues, but made two improvements - allowing table header text to wrap and taking viewport height into account.
